File: apps/api/app/models/deal_rooms.py
"""Collaborative Deal Room models for multi-party deal collaboration."""

from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Index, String, Text, UniqueConstraint
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import BaseModel, TimestampedModel


class DealRoom(BaseModel):
    """A secure, multi-party collaboration workspace tied to a project."""

    __tablename__ = "deal_rooms"

    org_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False, index=True)
    project_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("projects.id", ondelete="CASCADE"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    status: Mapped[str] = mapped_column(
        String(20), default="active", server_default="active"
    )  # active, closed, archived
    created_by: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    settings: Mapped[dict] = mapped_column(JSONB, default=dict)
    # {watermark: true, download_restricted: false, nda_required: true, expires_at: "2026-06-01"}


class DealRoomMember(BaseModel):
    """Member / participant in a deal room with role-based permissions."""

    __tablename__ = "deal_room_members"

    room_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("deal_rooms.id", ondelete="CASCADE"), nullable=False, index=True
    )
    user_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )
    email: Mapped[str | None] = mapped_column(String(255), nullable=True)
    role: Mapped[str] = mapped_column(String(20), nullable=False)  # owner, admin, member, viewer
    org_name: Mapped[str | None] = mapped_column(String(255), nullable=True)
    permissions: Mapped[dict] = mapped_column(JSONB, default=dict)
    # {can_upload: true, can_download: true, can_comment: true, can_view_financials: true}
    invited_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    joined_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    nda_signed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    invite_token: Mapped[str | None] = mapped_column(String(100), nullable=True, unique=True)

    __table_args__ = (
        Index("ix_deal_room_member_room_user", "room_id", "user_id"),
    )


class DealRoomDocument(BaseModel):
    """Document shared into a deal room (references existing data room document)."""

    __tablename__ = "deal_room_documents"

    room_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("deal_rooms.id", ondelete="CASCADE"), nullable=False, index=True
    )
    document_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    shared_by: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)

    __table_args__ = (
        UniqueConstraint("room_id", "document_id", name="uq_room_document"),
    )


class DealRoomMessage(BaseModel):
    """Threaded message in a deal room."""

    __tablename__ = "deal_room_messages"

    room_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("deal_rooms.id", ondelete="CASCADE"), nullable=False, index=True
    )
    user_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    parent_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("deal_room_messages.id", ondelete="SET NULL"), nullable=True
    )
    content: Mapped[str] = mapped_column(Text, nullable=False)
    mentions: Mapped[list] = mapped_column(JSONB, default=list)  # [user_id, ...]


class DealRoomActivity(TimestampedModel):
    """Immutable activity log for deal room audit trail."""

    __tablename__ = "deal_room_activities"

    room_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("deal_rooms.id", ondelete="CASCADE"), nullable=False, index=True
    )
    user_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    activity_type: Mapped[str] = mapped_column(String(50), nullable=False)
    # doc_uploaded, doc_viewed, doc_downloaded, comment, task_done, member_invited, message
    entity_type: Mapped[str | None] = mapped_column(String(50), nullable=True)
    entity_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)

    __table_args__ = (
        Index("ix_deal_room_activity_room_created", "room_id", "created_at"),
    )

File: apps/web/src/lib/deal-rooms.ts
"use client";

import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { api } from "@/lib/api";

// ── Types ─────────────────────────────────────────────────────────────────────

export interface DealRoom {
  id: string;
  name: string;
  project_id: string;
  status: string;
  created_by: string;
  settings: {
    nda_required?: boolean;
    download_restricted?: boolean;
    expires_at?: string | null;
  };
  members: DealRoomMember[];
  created_at: string;
}

export interface DealRoomMember {
  id: string;
  room_id: string;
  user_id: string | null;
  email: string | null;
  role: string;
  org_name: string | null;
  permissions: Record<string, boolean>;
  invited_at: string;
  joined_at: string | null;
  nda_signed_at: string | null;
}

export interface DealRoomMessage {
  id: string;
  room_id: string;
  user_id: string;
  parent_id: string | null;
  content: string;
  mentions: string[];
  created_at: string;
}

export interface DealRoomActivity {
  id: string;
  room_id: string;
  user_id: string;
  activity_type: string;
  entity_type: string | null;
  entity_id: string | null;
  description: string | null;
  created_at: string;
}

// ── Hooks ─────────────────────────────────────────────────────────────────────

export function useDealRooms() {
  return useQuery<DealRoom[]>({
    queryKey: ["deal-rooms"],
    queryFn: () => api.get("/deal-rooms/").then((r) => r.data),
  });
}

export function useDealRoom(roomId: string) {
  return useQuery<DealRoom>({
    queryKey: ["deal-room", roomId],
    queryFn: () => api.get(`/deal-rooms/${roomId}`).then((r) => r.data),
    enabled: !!roomId,
  });
}

export function useRoomMessages(roomId: string | null) {
  return useQuery<DealRoomMessage[]>({
    queryKey: ["deal-room-messages", roomId],
    queryFn: () =>
      api.get(`/deal-rooms/${roomId}/messages`).then((r) => r.data),
    enabled: !!roomId,
  });
}

export function useRoomActivity(roomId: string | null) {
  return useQuery<DealRoomActivity[]>({
    queryKey: ["deal-room-activity", roomId],
    queryFn: () =>
      api.get(`/deal-rooms/${roomId}/activity`).then((r) => r.data),
    enabled: !!roomId,
  });
}

export function useCreateRoom() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (body: Record<string, unknown>) =>
      api.post("/deal-rooms/", body).then((r) => r.data as DealRoom),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["deal-rooms"] }),
  });
}

export function useInviteMember(roomId: string) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (body: { email: string; role: string; org_name?: string }) =>
      api.post(`/deal-rooms/${roomId}/invite`, body).then((r) => r.data),
    onSuccess: () =>
      qc.invalidateQueries({ queryKey: ["deal-room", roomId] }),
  });
}

export function useSendMessage(roomId: string) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (content: string) =>
      api
        .post(`/deal-rooms/${roomId}/messages`, { content })
        .then((r) => r.data),
    onSuccess: () =>
      qc.invalidateQueries({ queryKey: ["deal-room-messages", roomId] }),
  });
}

export function useCloseRoom() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (roomId: string) =>
      api.post(`/deal-rooms/${roomId}/close`).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["deal-rooms"] }),
  });
}
